sphinx/environment/__init__.py

```diff
@@ -91,6 +92,17 @@
 }  # type: Dict[unicode, Union[bool, Callable]]
 
 
+@contextmanager
+def sphinx_smartquotes_action(env):
+    try:
+        original = SphinxSmartQuotes.smartquotes_action
+        SphinxSmartQuotes.smartquotes_action = env.config.smartquotes_action
+        yield
+        SphinxSmartQuotes.smartquotes_action = original
```

Comment:
This is not executed when the error has happened at yield block. Usually, it does not effect to the processing because `sphinx-build` crashes on the error. But the behavior might be changed in future. So it should be write back correctly using `final` clause.
How about this implementation?
```
if not hasattr(SphinxSmartQuotes, 'smartquotes_action'):
    # less than docutils-0.14
    yield
else:
    # docutils-0.14 or above
    try:
        original = SphinxSmartQuotes.smartquotes_action
        SphinxSmartQuotes.smartquotes_action = env.config.smartquotes_action
    finally:
        SphinxSmartQuotes.smartquotes_action = original
```

Reply:
+1, I guess I add a ``yield`` in the ``try`` ;-)